File: luau-obfuscator/src/analysis/scope.ts
/**
 * Semantic analysis: lexical scope resolution.
 *
 * Attaches a {@link Binding} to every declaration name and every `Name`
 * reference. References that resolve to no local binding are globals/builtins
 * and are left with `binding = null`. Upvalues need no special handling: a
 * closure referencing an outer local resolves to that same binding object, so a
 * later rename applied to the binding covers every use, at every depth.
 *
 * Correctly models Lua/Luau scoping rules:
 *  - a `local` is visible only after its declaration (RHS resolves first),
 *  - `local function f` is visible inside its own body (recursion),
 *  - loop variables are scoped to the loop body,
 *  - `repeat … until` can see locals declared in the body,
 *  - method bodies get an implicit, non-renamable `self`.
 */
import type {
  Binding, Block, Chunk, DeclName, Expr, FunctionExpr, Param, Stmt,
} from '../ast/nodes.js';

interface Scope {
  readonly vars: Map<string, Binding>;
  readonly parent: Scope | null;
}

export interface ScopeInfo {
  /** Every binding created, in declaration order. */
  readonly bindings: Binding[];
}

function newScope(parent: Scope | null): Scope {
  return { vars: new Map(), parent };
}

function declare(info: ScopeInfo, scope: Scope, decl: DeclName | Param, kind: Binding['kind']): Binding {
  const fixed = (decl as Param).implicit === true;
  const binding: Binding = { originalName: decl.name, newName: null, fixed, kind };
  decl.binding = binding;
  scope.vars.set(decl.name, binding);
  info.bindings.push(binding);
  return binding;
}

function lookup(scope: Scope, name: string): Binding | null {
  let s: Scope | null = scope;
  while (s) {
    const b = s.vars.get(name);
    if (b) return b;
    s = s.parent;
  }
  return null; // global / builtin
}
function resolve(scope: Scope, ref: { name: string; binding: Binding | null }): void {
  ref.binding = lookup(scope, ref.name);
}

function rExpr(info: ScopeInfo, e: Expr, scope: Scope): void {
  switch (e.type) {
    case 'Name': resolve(scope, e); break;
    case 'Paren': rExpr(info, e.expr, scope); break;
    case 'Field': rExpr(info, e.obj, scope); break;
    case 'Index': rExpr(info, e.obj, scope); rExpr(info, e.index, scope); break;
    case 'Call': rExpr(info, e.func, scope); e.args.forEach((a) => rExpr(info, a, scope)); break;
    case 'MethodCall': rExpr(info, e.obj, scope); e.args.forEach((a) => rExpr(info, a, scope)); break;
    case 'Binop': rExpr(info, e.left, scope); rExpr(info, e.right, scope); break;
    case 'Unop': rExpr(info, e.operand, scope); break;
    case 'Function': rFunction(info, e, scope); break;
    case 'Table':
      for (const f of e.fields) {
        if (f.kind === 'keyed') rExpr(info, f.key, scope);
        rExpr(info, f.value, scope);
      }
      break;
    default: break; // literals, Vararg
  }
}

function rFunction(info: ScopeInfo, fn: FunctionExpr, parent: Scope): void {
  const scope = newScope(parent);
  for (const p of fn.params) declare(info, scope, p, 'param');
  for (const st of fn.body) rStmt(info, st, scope);
}

function rBlock(info: ScopeInfo, stmts: Block, parent: Scope): void {
  const scope = newScope(parent);
  for (const st of stmts) rStmt(info, st, scope);
}

function rStmt(info: ScopeInfo, s: Stmt, scope: Scope): void {
  switch (s.type) {
    case 'LocalAssign':
      s.exprs.forEach((e) => rExpr(info, e, scope));
      s.names.forEach((n) => declare(info, scope, n, 'local'));
      break;
    case 'Assign':
      s.targets.forEach((e) => rExpr(info, e, scope));
      s.exprs.forEach((e) => rExpr(info, e, scope));
      break;
    case 'CompoundAssign':
      rExpr(info, s.target, scope);
      rExpr(info, s.value, scope);
      break;
    case 'CallStat': rExpr(info, s.call, scope); break;
    case 'Do': rBlock(info, s.body, scope); break;
    case 'While': rExpr(info, s.cond, scope); rBlock(info, s.body, scope); break;
    case 'Repeat': {
      // `until` sees body locals -> shared scope, cond resolved after body.
      const bs = newScope(scope);
      for (const st of s.body) rStmt(info, st, bs);
      rExpr(info, s.cond, bs);
      break;
    }
    case 'If':
      for (const c of s.clauses) { rExpr(info, c.cond, scope); rBlock(info, c.body, scope); }
      if (s.elseBody) rBlock(info, s.elseBody, scope);
      break;
    case 'NumericFor': {
      rExpr(info, s.start, scope); rExpr(info, s.stop, scope);
      if (s.step) rExpr(info, s.step, scope);
      const fs = newScope(scope);
      declare(info, fs, s.variable, 'loop');
      for (const st of s.body) rStmt(info, st, fs);
      break;
    }
    case 'GenericFor': {
      s.exprs.forEach((e) => rExpr(info, e, scope));
      const gs = newScope(scope);
      s.names.forEach((n) => declare(info, gs, n, 'loop'));
      for (const st of s.body) rStmt(info, st, gs);
      break;
    }
    case 'FunctionDecl':
      if (s.isLocal) {
        declare(info, scope, s.base, 'function'); // visible in its own body
        rFunction(info, s.func, scope);
      } else {
        resolve(scope, { name: s.base.name, get binding() { return null; }, set binding(b) { s.base.binding = b ?? undefined; } } as unknown as { name: string; binding: Binding | null });
        rFunction(info, s.func, scope);
      }
      break;
    case 'Return': s.exprs.forEach((e) => rExpr(info, e, scope)); break;
    default: break; // Break, Continue, Goto, Label
  }
}

/** Resolve scopes across a chunk; returns every binding for the rename pass. */
export function analyzeScopes(chunk: Chunk): ScopeInfo {
  const info: ScopeInfo = { bindings: [] };
  const top = newScope(null);
  for (const st of chunk.body) rStmt(info, st, top);
  return info;
}

File: luau-vm/src/scope.lua
-- luau-vm/src/scope.lua
-- Capture analysis. Resolves every Name to its declaration and marks a
-- declaration `captured` when it is referenced from a nested function. The
-- compiler uses `captured` to decide whether a local lives in a cell (box) so
-- closures can share it as an upvalue.
--
-- Declaration objects are attached to the AST and reused by the compiler, which
-- fills in `owner`/`reg` during code generation.

local Scope = {}

local function newScope(parent, fdepth) return { vars = {}, parent = parent, fdepth = fdepth } end
local function declare(scope, name)
  local d = { name = name, captured = false, fdepth = scope.fdepth }
  scope.vars[name] = d
  return d
end
local function lookup(scope, name)
  local s = scope
  while s do local d = s.vars[name]; if d then return d end; s = s.parent end
  return nil
end

local rExpr, rStmt, rBlock, rFunc

function Scope.analyze(chunk)
  local top = newScope(nil, 0)
  for _, s in ipairs(chunk.body) do rStmt(s, top) end
  return chunk
end

function rBlock(stmts, parent)
  local scope = newScope(parent, parent.fdepth)
  for _, s in ipairs(stmts) do rStmt(s, scope) end
end

function rFunc(fn, parent)
  local scope = newScope(parent, parent.fdepth + 1)
  fn.paramDecls = {}
  for i, p in ipairs(fn.params) do fn.paramDecls[i] = declare(scope, p) end
  for _, s in ipairs(fn.body) do rStmt(s, scope) end
end

local function ref(node, scope)
  local d = lookup(scope, node.name)
  if d then
    node.decl = d
    if d.fdepth ~= scope.fdepth then d.captured = true end
  else
    node.decl = nil -- global
  end
end

function rExpr(e, scope)
  local k = e.k
  if k == 'Name' then ref(e, scope)
  elseif k == 'Paren' then rExpr(e.expr, scope)
  elseif k == 'Field' then rExpr(e.obj, scope)
  elseif k == 'Index' then rExpr(e.obj, scope); rExpr(e.index, scope)
  elseif k == 'CallE' then rExpr(e.func, scope); for _, a in ipairs(e.args) do rExpr(a, scope) end
  elseif k == 'Method' then rExpr(e.obj, scope); for _, a in ipairs(e.args) do rExpr(a, scope) end
  elseif k == 'Binop' then rExpr(e.left, scope); rExpr(e.right, scope)
  elseif k == 'IfExpr' then
    for _, c in ipairs(e.clauses) do rExpr(c.cond, scope); rExpr(c.value, scope) end
    rExpr(e.elseValue, scope)
  elseif k == 'Unop' then rExpr(e.operand, scope)
  elseif k == 'Function' then rFunc(e, scope)
  elseif k == 'Table' then
    for _, f in ipairs(e.fields) do
      if f.kind == 'keyed' then rExpr(f.key, scope) end
      rExpr(f.value, scope)
    end
  end
end

function rStmt(s, scope)
  local k = s.k
  if k == 'Local' then
    for _, e in ipairs(s.exprs) do rExpr(e, scope) end
    s.decls = {}
    for i, nm in ipairs(s.names) do s.decls[i] = declare(scope, nm) end
  elseif k == 'Assign' then
    for _, t in ipairs(s.targets) do rExpr(t, scope) end
    for _, e in ipairs(s.exprs) do rExpr(e, scope) end
  elseif k == 'CallStat' then rExpr(s.call, scope)
  elseif k == 'Do' then rBlock(s.body, scope)
  elseif k == 'While' then rExpr(s.cond, scope); rBlock(s.body, scope)
  elseif k == 'Repeat' then
    local inner = newScope(scope, scope.fdepth)
    for _, st in ipairs(s.body) do rStmt(st, inner) end
    rExpr(s.cond, inner)
  elseif k == 'If' then
    for _, c in ipairs(s.clauses) do rExpr(c.cond, scope); rBlock(c.body, scope) end
    if s.elseBody then rBlock(s.elseBody, scope) end
  elseif k == 'NumFor' then
    rExpr(s.start, scope); rExpr(s.stop, scope); if s.step then rExpr(s.step, scope) end
    local inner = newScope(scope, scope.fdepth)
    s.varDecl = declare(inner, s.var)
    for _, st in ipairs(s.body) do rStmt(st, inner) end
  elseif k == 'GenFor' then
    for _, e in ipairs(s.exprs) do rExpr(e, scope) end
    local inner = newScope(scope, scope.fdepth)
    s.decls = {}
    for i, nm in ipairs(s.names) do s.decls[i] = declare(inner, nm) end
    for _, st in ipairs(s.body) do rStmt(st, inner) end
  elseif k == 'FuncDecl' then
    -- base is a reference (assignment target); resolve it
    local nameNode = { k = 'Name', name = s.base, line = s.line }
    ref(nameNode, scope)
    s.baseDecl = nameNode.decl
    rFunc(s.func, scope)
  elseif k == 'LocalFunc' then
    s.decl = declare(scope, s.name) -- visible in its own body (recursion)
    rFunc(s.func, scope)
  elseif k == 'Return' then
    for _, e in ipairs(s.exprs) do rExpr(e, scope) end
  end
  -- Break / Goto / Label: nothing
end

return Scope
